File: 2021/Round 2/Matrygons.java
import java.util.*;
import java.io.*;

public class Matrygons {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(System.out);
        int numTests = Integer.parseInt(br.readLine());

        for (int test = 1; test <= numTests; test++) {
            int N = Integer.parseInt(br.readLine());
            pw.println(String.format("Case #%d: %d", test, getMatrygons(N)));
        }
        pw.flush();
        pw.close();
        br.close();
    }

    static int getMatrygons(int N) {
        int[] K = new int[N + 1];
        int ans = 1;

        for (int i = 3; i < N; i++) {
            if (N % i == 0) {
                ans = Math.max(ans, getK(K, N / i));
            }
        }
        // System.out.println(Arrays.toString(K));
        return ans;
    }

    static int getK(int[] K, int n) {
        if (n <= 2) {
            K[n] = 1;
        } else if (K[n] == 0) {
            K[n] = 2;
            for (int i = 2; i <= n; i++) {
                if ((n - 1) % i == 0) {
                    K[n] = Math.max(K[n], getK(K, (n - 1) / i) + 1);
                }
            }
        }
        // System.out.println(String.format("Input N = %d, K = %d", n, K[n]));
        return K[n];
    }
}

File: 2021/Round 2/matrygons.py
import sys


def get_K(K, n):
    if n <= 2:
        K[n] = 1
    elif n not in K:
        K[n] = 2  # when n is prime
        for i in range(2, n + 1):
            if (n - 1) % i == 0:
                K[n] = max(K[n], get_K(K, (n - 1) // i) + 1)
    return K[n]


def get_matrygons(N):
    K = {}
    ans = 1

    for i in range(3, N):
        if N % i == 0:
            ans = max(ans, get_K(K, N//i))
    return ans


def main():
    num_tests = int(input())

    for t in range(1, num_tests + 1):
        N = int(input())
        print(f'Case #{t}: {get_matrygons(N)}')

    sys.stdout.flush()


if __name__ == '__main__':
    main()
